scripts/qemu.py

```diff
@@ -125,38 +134,46 @@ def _post_shutdown(self):
             self._remove_if_exists(self._monitor_address)
         self._remove_if_exists(self._qemu_log_path)
 
-    def launch(self):
-        '''Launch the VM and establish a QMP connection'''
-        devnull = open('/dev/null', 'rb')
-        qemulog = open(self._qemu_log_path, 'wb')
+    def launch(self, timeout=0.1):
         try:
-            self._pre_launch()
-            args = self._wrapper + [self._binary] + self._base_args() + self._args
-            self._popen = subprocess.Popen(args, stdin=devnull, stdout=qemulog,
-                                           stderr=subprocess.STDOUT, shell=False)
-            self._post_launch()
+            self._launch(timeout)
         except:
-            if self.is_running():
-                self._popen.kill()
-                self._popen.wait()
-            self._load_io_log()
-            self._post_shutdown()
+            self.shutdown()
             raise
 
+    def _launch(self, timeout):
+        '''Launch the VM and establish a QMP connection'''
+        devnull = open('/dev/null', 'rb')
+        qemulog = open(self._qemu_log_path, 'wb')
+        args = self._wrapper + [self._binary] + self._base_args() + self.args
+        self._pre_launch()
+        self._popen = subprocess.Popen(args, stdin=devnull, stdout=qemulog,
+                                       stderr=subprocess.STDOUT, shell=False)
+
+        start = time.time()
+        end = start + timeout
+        while time.time() < end:
+            time.sleep(0.1)
+            if not self.is_running():
+                self._load_io_log()
+                raise QEMULaunchError('Error launching QEMU.\n'
+                                      'Command: \n%s\nOutput: \n%s\n' %
+                                      (' '.join(args), ''.join(self._iolog)))
```

Comment:
I don't really like the `timeout` as it's not a timeout, but more like `is_running_timeslot`. Also by default it fails with qemu executions which are meant to finish immediately (you can avoid that by manually setting `launch(-1)`, but it changes the default) and it makes every test if 0.1s (or timeout) longer even though you probably find out the machine is not running on interaction (monitor command, ...).

Also it's unlikely that the process would die and re-execute so you can get (more-less) the same behavior (for non-malfunctioning execution) by simple `time.sleep(time.time() + timeout); self.is_running()`.

If you really want feature like this I'd support `watch_period` argument to `is_running` and allow users to call it from tests:

```python
vm.launch()
if not vm.is_running(10, 0.1):
    ...
```